Migrate Merchandise Return - Order Detail List

JevgenijVisockij opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates "merchandise return" product list to Symfony
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/24421
| How to test?  | See below

## How to test

1. Enable merchandise returns option
2. Create some merchandise returns.
3. Enable the Merchandise return page in experimental features
4. Go to `admin-dev/index.php/sell/customer-service/merchandise-return/` which is the list. Click on any of the existing returns to open the page.
5. You should be able to see list of products in that merchandise return. They should be deletable using delete action or bulk action. Merchandise return must always have at least one product left.

JevgenijVisockij committed:
Cs fixer fixes

JevgenijVisockij committed:
Updated bulk order detail delete handler to use repo

JevgenijVisockij committed:
Updated order return pdf printer to use const

JevgenijVisockij committed:
Fixed order return pdf printing

JevgenijVisockij committed:
Fixed merchandise return deletion

JevgenijVisockij committed:
CS fixer fix

JevgenijVisockij committed:
Changed order return to use function from repository

# Conflicts:
#	src/Adapter/OrderReturn/AbstractOrderReturnHandler.php

JevgenijVisockij committed:
PHP stan fixes

JevgenijVisockij committed:
CS fixer and stan fixes

JevgenijVisockij committed:
PHP stan fix

JevgenijVisockij committed:
Yaml lint fix

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/adapter/pdf/generator.yml:
```diff
@@ -32,3 +32,8 @@ services:
     class: PrestaShop\PrestaShop\Adapter\PDF\InvoicePdfGenerator
+
+  PrestaShop\PrestaShop\Adapter\PDF\OrderReturnPdfGenerator:
+    autowire: true
```
The purpose of autowire is to avoid setting manually all the dependencies for each service. You can remove them and only specify the needed dependencies.

This applies to every configuration file is this PR.

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml:
```diff
@@ -456,3 +456,11 @@ services:
     arguments:
       - '@prestashop.core.query.doctrine_search_criteria_applicator'
+
+  PrestaShop\PrestaShop\Core\Grid\Query\OrderReturnProductsQueryBuilder:
```
There is a problem with the namespace here, the `OrderReturnProductsQueryBuilder` should be inside the 
PrestaShop\PrestaShop\Core\Grid\ namespace i think. But maybe in the Prestashop bundle.

To be confirmed by someone else.

JevgenijVisockij replied:
I don't quite get why it shouldn't be in  PrestaShop\PrestaShop\Core\Grid\Query. That's where the file is.

FabienPapet commented on src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php:
```diff
@@ -151,13 +258,44 @@ private function getErrorMessages(): array
+
+        // When using legacy generator,
+        // we want to be sure that displaying PDF is the last thing this controller will do
+        die();
```
👀

FabienPapet replied:
There is a streamed response that you can use for this.

FabienPapet commented on src/Adapter/OrderReturn/QueryHandler/GetOrderDetailCustomizationsHandler.php:
```diff
@@ -0,0 +1,88 @@
+     */
+    public function handle(GetOrderDetailCustomizations $query): ?OrderDetailCustomizations
+    {
+        $orderDetail = new OrderDetail($query->getOrderDetailId()->getValue());
```
Please create repositories for this, with their interfaces in the core, to avoid code duplication. And to ease refactoring later. 

Interfaces should return model objects (inside the core namespace). And their implementation should be placed inside the Adapter namespace.


`new Order``Product::getAllCustomizedDatas`

FabienPapet commented on src/Adapter/PDF/OrderReturnPdfGenerator.php:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function generatePDF(array $orderReturnId)
```
Why rely on this interface ? This interface look very wrong because it takes only one argument and is expecting an array... 

I think we should create a better interface here.
WDYT ?

JevgenijVisockij replied:
The whole src/Adapter/PDF needs to be refactored as a whole, but as you said this PR is already big enough so it's not a good idea to do it here. That's why for this I am following already established conventions.

FabienPapet replied:
In this case, you can introduce a new interface in a dedicated PR, and rebase your work onto this one. We must not add any temporary code that we'll need to refactor later. It does not help with the migration.

Then you can rebase your work onto this one.

JevgenijVisockij replied:
IMO the PDF rework shouldn't block this PR, since the amount of code used here is rather small

FabienPapet replied:
It is blocking, because the interface is wrong, and we should not use it anymore.
Because the signature contains an array, and we don't use an array, but an object. 

If you don't want to add a new interface here, I can be ok with this, but  don't use the old one.

And IMO, if it's small, then it's easy to add.

jolelievre replied:
I agree the interface is not right, but it's not the job of this PR to fix/improve it So not blocking here we can't fix everything everywhere everytime we notice something is wrong or we never move forward 😅

jolelievre replied:
Actually, it's not only the interface but the way PDF are generated (I believe it directly outputs on the stream) which is incorrect but we don't have a component to replace this yet So we should create a new component to generate PDF properly and replace the current usages. So using the old incorrect interface here is actually interesting for the future refacto as it will allow us to find all the use cases more easily

FabienPapet commented on src/Adapter/PDF/OrderReturnPdfGenerator.php:
```diff
@@ -0,0 +1,82 @@
+        }
+
+        $orderReturnId = reset($orderReturnId);
+        $orderReturn = new OrderReturn((int) $orderReturnId);
```
This should also be part of a repository, that can throw those exceptions.

FabienPapet commented on src/Core/Domain/OrderReturn/QueryResult/OrderDetailCustomization.php:
```diff
@@ -0,0 +1,105 @@
+        $this->name = $name;
+        $this->value = $value;
+        if (Product::CUSTOMIZE_FILE === $this->type) {
+            $this->image = _THEME_PROD_PIC_DIR_ . $this->value . '_small';
```
Please don't use a constant here. This check should be done outside of this PR.

JevgenijVisockij replied:
Changed to use OrderDetailCustomizationForViewing after close look

FabienPapet commented on src/Core/Grid/Column/Type/OrderReturnCustomizationColumn.php:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Class OrderReturnColumn defines column that renders customization data for order return products
+ */
+class OrderReturnCustomizationColumn extends AbstractColumn
```
Looks like this column is just a configured column, IIRC we have a configurable column that can be used to avoid creating files like this one.

ping @jolelievre @mflasquin

jolelievre replied:
@FabienPapet which configurable column are you refering to? As far as I know we don't have configurable columns 🤔

However, what you could do @JevgenijVisockij, since this column has no intelligence and no specific options you can use a basic `DataColumn` (with `field = customizations`) and then you can override the column template for this specific column. The grid system has an automatic template override mechanism based on:
- grid ID
- column ID
- column type

So in the column templates folder https://github.com/PrestaShop/PrestaShop/tree/8.1.x/src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/Content You can add a `order_return_products_customizations_data.html.twig` that will replace the default twig template and you can handle your custom template here

jolelievre replied:
Also I believe what @FabienPapet suggested, but I don't think it exists yet, is to have a column in which you could simply specify a twig template to use. You can also create such a column which would come handy for modules and even for us because we can store our custom twig templates in more specified folders instead of increasing the grid default one forever.

jolelievre replied:
So you could introduce a `TemplateColumn` that extends `AbstractColumn` and adds a required option `template_path` so you can basically use ut this way:

```php
            ->add(
                (new TemplateColumn('customizations'))
                    ->setName($this->trans('Customizations', [], 'Admin.Global'))
                    ->setOptions([
                        'template_path' => '@PrestaShop/Admin/Sell/CustomerService/OrderReturn/customizations_column.html.twig',
                    ])
            )
```

I don't think it needs any more options for now, since inside it you can pretty much do any custom hard coded data retrieval, and if needed it can always be added later.

jolelievre replied:
Oh and in the generic layout `src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/Content/custom_template.html.twig` you will only have to include the specified template path:

```twig
{{ include(column.options.template_path) }}
```
or
```twig
{{ include(column.options.template_path, {'record': record, 'column': column, 'grid': grid}) }}
```
because I don't know if the variables will be accessible out of the box or not, you need to try.

JevgenijVisockij replied:
I will go with TemplateColumn, IMO that makes the most sense. Thanks!

FabienPapet replied:
That would make columns classe more reusable, thank you @jolelievre for the explanation. That's  what I was thinking.

FabienPapet commented on src/Core/Grid/Query/OrderReturnProductsQueryBuilder.php:
```diff
@@ -0,0 +1,175 @@
+        Connection $connection,
+        string $dbPrefix,
+        DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator,
+        int $contextLanguageId
```
A repository should not be aware of the context, so this dependency should be removed. And the language Id, added to the methods it is used.

JevgenijVisockij replied:
This is query builder for the grid system, it's very common to use contextLanguageId, passing langId to methods here would require big changes to grid logic.

zuk3975 replied:
actually languageId could be built in searchCriteria, but that might be a bit too much work for this PR which is being dragged for months, so Im ok if we improve it in other PR

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/bundle/event_listener.yml:
```diff
@@ -125,3 +125,12 @@ services:
+  PrestaShopBundle\EventListener\FilterOrderReturnProductsSearchCriteriaListener:
+    autowire: true
+    arguments:
+      - '@request_stack'
```
```suggestion

```

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/grid/grid_filter_form_factory.yml:
```diff
@@ -35,3 +35,10 @@ services:
+
+  PrestaShop\PrestaShop\Core\Grid\Filter\OrderReturnProductsFilterFormFactory:
+    autowire: true
+    arguments:
+      - '@prestashop.core.grid.filter.form_factory'
+      - '@router'
+      - '@request_stack'
```
```suggestion
    arguments:
      - '@prestashop.core.grid.filter.form_factory'
```

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/grid/grid_factory.yml:
```diff
@@ -328,6 +328,14 @@ services:
 
+  PrestaShop\Core\Grid\Factory\OrderReturnProducts:
+    class: 'PrestaShop\PrestaShop\Core\Grid\GridFactory'
+    arguments:
+      - '@PrestaShop\PrestaShop\Core\Grid\Definition\Factory\OrderReturnProductsGridDefinitionFactory'
+      - '@PrestaShop\Core\Grid\Data\Factory\OrderReturnProductsDecorator'
+      - '@PrestaShop\PrestaShop\Core\Grid\Filter\OrderReturnProductsFilterFormFactory'
+      - '@prestashop.core.hook.dispatcher'
```
```suggestion
    autowire: true
```

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/grid/grid_definition_factory.yml:
```diff
@@ -266,6 +266,13 @@ services:
+    autowire: true
+    parent: 'prestashop.core.grid.definition.factory.abstract_grid_definition'
+    public: true
+    arguments:
+      - '@request_stack'
```
```suggestion
```

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml:
```diff
@@ -361,6 +361,21 @@ services:
+      - 'order_return_products'
+
+  PrestaShop\Core\Grid\Data\Factory\OrderReturnProductsDecorator:
+    class: 'PrestaShop\PrestaShop\Core\Grid\Data\Factory\OrderReturnProductsGridDataFactory'
```
```suggestion
```

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml:
```diff
@@ -361,6 +361,21 @@ services:
+    class: '%prestashop.core.grid.data.factory.doctrine_grid_data_factory%'
+    autowire: true
+    arguments:
+      - '@PrestaShop\PrestaShop\Core\Grid\Query\OrderReturnProductsQueryBuilder'
+      - '@prestashop.core.hook.dispatcher'
+      - '@prestashop.core.grid.query.doctrine_query_parser'
+      - 'order_return_products'
```
```suggestion
      $theVariableName: 'order_return_products'
```

FabienPapet commented on src/Adapter/PDF/OrderReturnPdfGenerator.php:
```diff
@@ -0,0 +1,82 @@
+     */
+    public function generatePDF(array $orderReturnId)
+    {
+        if (count($orderReturnId) !== 1) {
+            throw new CoreException(sprintf('"%s" supports generating order return for single order only.', get_class($this)));
+        }
```
Useless with the correct interface usage

FabienPapet commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -86,4 +109,121 @@ public function update(OrderReturn $orderReturn): void
+        }
+    }
+
+    public function getOrderReturnDetailByOrderDetailId(int $orderDetailId): ?OrderReturnDetail
```
You can throw an exception if not found. That'll make the development easier as we will not need to check null vars.

FabienPapet commented on src/Adapter/OrderReturn/CommandHandler/DeleteProductFromOrderReturnHandler.php:
```diff
@@ -0,0 +1,67 @@
+<?php
```
There's no legacy here, so this should be placed inside the core.

JevgenijVisockij replied:
Moved

FabienPapet commented on src/Adapter/OrderReturn/CommandHandler/BulkDeleteProductFromOrderReturnHandler.php:
```diff
@@ -0,0 +1,131 @@
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\OrderReturn\CommandHandler;
```
There's no legacy here, so this should be placed inside the core.

JevgenijVisockij replied:
Moved

FabienPapet commented on src/Adapter/OrderReturn/CommandHandler/BulkDeleteProductFromOrderReturnHandler.php:
```diff
@@ -0,0 +1,131 @@
+<?php
```
I can't add comment on empty file, but src/Adapter/OrderReturn/AbstractOrderReturnHandler.php is empty

JevgenijVisockij replied:
Deleted it

FabienPapet commented:
This PR was 2,5k LOC, can you make smaller ones for easier reviewing ? :D

JevgenijVisockij committed:
Added customizations to repository, changed PDF generator to use repository

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/config/services/bundle/event_listener.yml

Co-authored-by: Fabien Papet <FabienPapet@users.noreply.github.com>

JevgenijVisockij committed:
Apply suggestions from code review

Autowiring suggestions

Co-authored-by: Fabien Papet <FabienPapet@users.noreply.github.com>

JevgenijVisockij committed:
Fixed to yml failes for autowiring

JevgenijVisockij committed:
Deleting uneeded file

JevgenijVisockij committed:
Moved handlers to core since they don't have legacy code

JevgenijVisockij committed:
Changed repository to throw exception instead of returning null

JevgenijVisockij committed:
Changed to use OrderProductCustomizationForViewing

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/adapter/pdf/generator.yml:
```diff
@@ -32,3 +32,8 @@ services:
     class: PrestaShop\PrestaShop\Adapter\PDF\InvoicePdfGenerator
+
+  PrestaShop\PrestaShop\Adapter\PDF\OrderReturnPdfGenerator:
+    arguments:
```
```suggestion
    public: false
    arguments:
```

JevgenijVisockij replied:
Could you elaborate why this should be public:false?

FabienPapet replied:
Because all services are meant to be private by default.  Only some services can be public (like controllers for example) but these are exceptions.

FabienPapet replied:
Because all services are meant to be private by default.  Only some services can be public (like controllers for example) but these are exceptions.

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/bundle/event_listener.yml:
```diff
@@ -125,3 +125,11 @@ services:
       - { name: kernel.event_listener, method: onKernelRequest }
+
+  PrestaShopBundle\EventListener\FilterOrderReturnProductsSearchCriteriaListener:
+    autowire: true
```
```suggestion
    public: false
    autowire: true
```

FabienPapet commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -86,4 +110,119 @@ public function update(OrderReturn $orderReturn): void
+                    DeleteOrderReturnProductException::UNEXPECTED_ERROR
+                );
+            }
+        } catch (PrestaShopException $e) {
```
Maybe you can use another Exception than the legacy one. So we don't add more legacy usages.

JevgenijVisockij replied:
Well this catch here is in case there is a legacy exception to throw a core exception in this case DeleteOrderReturnProductException

FabienPapet replied:
ok

JevgenijVisockij committed:
Moved Merchandise return twig files to relevant folder, added template column

JevgenijVisockij committed:
Autowiring rest of the services

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/config/services/bundle/event_listener.yml

Making event_listener public false

Co-authored-by: Fabien Papet <FabienPapet@users.noreply.github.com>

JevgenijVisockij committed:
Yaml linter fix

JevgenijVisockij commented:
> This PR was 2,5k LOC, can you make smaller ones for easier reviewing next time ? :D

Will do my best, this migration was already split into 2 parts, but yeah probably could have separated even more

JevgenijVisockij committed:
Changed to use streamed response

JevgenijVisockij committed:
fixed phpdoc comment

JevgenijVisockij commented:
@FabienPapet thanks a lot for reviews. Left couple smaller questions and implemented the rest aside for Pdf Generator changes.

JevgenijVisockij committed:
Making pdf generator public: false

FabienPapet commented on src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php:
```diff
@@ -27,17 +27,27 @@
 namespace PrestaShopBundle\Controller\Admin\Sell\CustomerService;
 
 use Exception;
+use PrestaShop\PrestaShop\Adapter\PDF\OrderReturnPdfGenerator;
```
The `OrderReturnPdfGenerator` should have an interface inside the core and being inject here instead of the adapter.

The purpose is to have interfaces inside the core, and their implementations inside the adapter namespace. So we can migrate every service from adapters to the core.

FabienPapet commented on src/Adapter/PDF/OrderReturnPdfGenerator.php:
```diff
@@ -0,0 +1,90 @@
+     * @param TranslatorInterface $translator
+     * @param OrderReturnRepository $orderReturnRepository
+     */
+    public function __construct(TranslatorInterface $translator, OrderReturnRepository $orderReturnRepository)
```
The injected repository should be behind an interface, so we can easily replace it.

JevgenijVisockij replied:
I am not sure about repository interface because main function get for both OrderReturnRepository and OrderRepository returns legacy objectModel (Order and OrderReturn). When we are replacing repositories we will probably refactor object model as well so such interfaces would have to be refactored too.

jolelievre replied:
We don't have the interfaces available yet, so not blocking for now

FabienPapet commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -28,13 +28,22 @@
 use PrestaShop\PrestaShop\Core\Repository\AbstractObjectModelRepository;
+use PrestaShopException;
 
 class OrderReturnRepository extends AbstractObjectModelRepository
```
Missing interface here

jolelievre replied:
The interfaces haven't been finalized yet, we can't block this PR because of this

JevgenijVisockij committed:
fixed from private to public

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/grid/grid_filter_form_factory.yml:
```diff
@@ -35,3 +35,8 @@ services:
+  PrestaShop\PrestaShop\Core\Grid\Filter\OrderReturnProductsFilterFormFactory:
+    autowire: true
+    arguments:
+      - '@prestashop.core.grid.filter.form_factory'
```
```suggestion
```

FabienPapet commented on src/Core/Domain/OrderReturn/Query/GetOrderDetailCustomizations.php:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @throws OrderException
+     */
+    public function __construct(int $orderDetailId)
```
I think this should have the context language set as well, so we don't rely on the context

zuk3975 commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -71,4 +77,37 @@ public function get(OrderId $orderId): Order
+    {
+        $orderDetail = new OrderDetail($detailId->getValue());
+        $order = new Order($orderDetail->id_order);
+        $customizations = [];
```
just a nitpick if you have other things to fix (not blocking)
This var will end up unused if (!is_array($customizedDatas)).
So ideally it should be moved before foreach loop to avoid initializing it for no wreason

zuk3975 commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -86,4 +110,122 @@ public function update(OrderReturn $orderReturn): void
+     * @param OrderReturnId $orderReturnId
+     * @param Order $order
+     *
+     * @return array<OrderReturnDetail>
```
Im not sure current definition is legit, I cannot find it in psalm docs. (why I doubt is because usually the first item of the annotation is the key and the second one is value, so it should look like array<int, OrderReturnDetail>). Anyway if you know this way is legit, please just provide me some link/argument so I know for future too :smile: 
```suggestion
     * @return OrderReturnDetail[]
```

jolelievre replied:
Agreed, although PHPStan doesn't seem to complain so maybe it's fine

zuk3975 commented on src/Core/Domain/OrderReturn/QueryResult/OrderDetailCustomizations.php:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * @param OrderProductCustomizationForViewing $orderDetailCustomization
+     */
+    private function addOrderDetailCustomization(OrderProductCustomizationForViewing $orderDetailCustomization): void
+    {
+        $this->orderDetailCustomizations[] = $orderDetailCustomization;
+    }
```
kind of redundant method.
I would put the whole loop into this method to make more use of it.
```suggestion
    private function setOrderDetailCustomizations(array $orderDetailCustomizations): void
    {
        foreach ($orderDetailCustomizations as $orderDetailCustomization) {
            $this->orderDetailCustomizations[] = $orderDetailCustomization;
        }
    }
```

zuk3975 commented on src/Core/Grid/Definition/Factory/OrderReturnProductsGridDefinitionFactory.php:
```diff
@@ -0,0 +1,255 @@
+     *
+     * @throws OrderReturnProductException
+     */
+    private function setOrderReturnId(RequestStack $requestStack): void
```
private method should go down bellow protected ones

zuk3975 replied:
Btw is real bad practice to depend on request here. I know there is currently no other option, but I think we should alter the design of grid definition factory to allow adding some runtime parameters for such cases.
E.g. introduce optional $options param into GridFactory->getGrid($filters, array $options). This would also require same for GridDefinitionFactory->getDefinition($options = []). But I think its doable without BC breaks this way.
In the end in controller then you could call it like: 
```
        $gridFactory->getGrid(
            $filters,
            ['orderReturnId' => $request->request->get('orderReturnId')]
        );
```

Anyway I think this deserves a dedicated PR. So for now It can stay this way
Wdyt @JevgenijVisockij, @jolelievre?

zuk3975 replied:
we've with @jolelievre figured a way to integrate some options (or rather filters) during runtime in a much better way.
You should create dedicated GridFactory and a dedicated filters builder, to pass orderReturnId into $filters.
See example in this PR with feature value https://github.com/PrestaShop/PrestaShop/pull/32334/files#diff-45667b705fa4b87859bc55a90391bec7aa547b2e72dba20d589b0d9dc6d68047R112

zuk3975 commented on src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php:
```diff
@@ -119,14 +135,106 @@ public function editAction(int $orderReturnId, Request $request): Response
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages()));
+        }
+
+        return $this->redirectToRoute(
+            'admin_order_returns_edit',
+            [
+                'orderReturnId' => $orderReturnId,
+            ]
+        );
```
just optional a style suggestion, I think its ok to do it in one line when there are no more params. 
```suggestion
        return $this->redirectToRoute('admin_order_returns_edit', ['orderReturnId' => $orderReturnId]);
```

zuk3975 replied:
same can be applied in other places. But not blocking, has no practical impact anyways

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/adapter/order_return.yml:
```diff
@@ -21,10 +21,47 @@ services:
+    tags:
+      - name: tactician.handler
+        command: 'PrestaShop\PrestaShop\Core\Domain\OrderReturn\Command\BulkDeleteProductFromOrderReturnCommand'
+    public: false
```
you correctly added public:false here, but all the services above are missing this :arrow_up:

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml:
```diff
@@ -471,3 +471,11 @@ services:
+  PrestaShop\PrestaShop\Core\Grid\Query\OrderReturnProductsQueryBuilder:
+    autowire: true
+    parent: 'prestashop.core.grid.abstract_query_builder'
+    public: true
```
shouldnt be false?
```suggestion
    public: false
```

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/core/grid/grid_definition_factory.yml:
```diff
@@ -266,6 +266,11 @@ services:
+  PrestaShop\PrestaShop\Core\Grid\Definition\Factory\OrderReturnProductsGridDefinitionFactory:
+    autowire: true
+    parent: 'prestashop.core.grid.definition.factory.abstract_grid_definition'
+    public: true
```
```suggestion
    public: false
```

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/core/grid/grid_filter_form_factory.yml:
```diff
@@ -35,3 +35,8 @@ services:
       - 'admin_credit_slips_search'
+
+  PrestaShop\PrestaShop\Core\Grid\Filter\OrderReturnProductsFilterFormFactory:
+    autowire: true
```
```suggestion
    autowire: true
    public: false
```

zuk3975 commented:
big PR, project evolves and new changes are requested every time :hot_face:
Hopefully we can finish it up soon :smile:

Hlavtox commented on src/PrestaShopBundle/Resources/views/Admin/Sell/CustomerService/MerchandiseReturn/Blocks/form.html.twig:
```diff
@@ -49,3 +63,11 @@
 {{ form_end(orderReturnForm) }}
+
+{% block order_returns_products %}
+  <div class="row">
```
`<div class="row"><div class="col-12">` not needed

Hlavtox commented on src/PrestaShopBundle/Resources/views/Admin/Sell/CustomerService/MerchandiseReturn/customizations_column.html.twig:
```diff
@@ -0,0 +1,45 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ *#}
+
+<div class="text-nowrap">
+  <div class="text-nowrap">
+    {% for customizations in [record['customizations']] %}
+      {% if customizations is  not null %}
+        {% for customization in customizations.orderDetailCustomizations %}
+          <div class="row">
+            <span class="col-lg-12"><strong>{{ customization.name }}</strong></span>
+            <div class="col-lg-12">
+              {% if customization.image is not null %}
+                <img class="img-thumbnail" src="{{ customization.image }}" alt="{{ customization.value }}">
+              {% else %}
+                {{ customization.value }}
+              {% endif %}
+            </div>
+          </div>
+        {% endfor %}
+      {% endif %}
+    {% endfor %}
+  </div>
+</div>
```
```suggestion
<div>
  {% for customizations in [record['customizations']] %}
    {% if customizations is  not null %}
      {% for customization in customizations.orderDetailCustomizations %}
        <div><strong>{{ customization.name }}</strong></div>
        <div>
          {% if customization.image is not null %}
            <img class="img-thumbnail" src="{{ customization.image }}" alt="{{ customization.value }}">
          {% else %}
            {{ customization.value }}
          {% endif %}
        </div>
      {% endfor %}
    {% endif %}
  {% endfor %}
</div>
```

This should be enough ;-)

jolelievre commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -86,4 +110,122 @@ public function update(OrderReturn $orderReturn): void
+    ): void {
+        $orderReturn = $this->get($orderReturnId);
+
+        if ((int) ($orderReturn->countProduct()) <= 1) {
```
I'm not sure to understand the domain rule here 🤔 The order return must have more than one product to be deletable? So any return with 2, 67, 1000 is fine but 0 or 1 no This is weird no?

jolelievre replied:
From what I understand in the bulk delete as well, you can delete the order return details from an order return, but one must always remain, right?

JevgenijVisockij replied:
Yes exactly, you can't delete last product in order return, because then that order return won't make sense.

I could add comment to clarify, but the exception already says the reason so I am not sure if it's needed. I could change to = 1, to be more clear.

jolelievre commented on src/Core/Domain/Order/ValueObject/OrderDetailId.php:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @throws OrderException
+     */
+    public function __construct($orderDetailId)
```
```suggestion
    public function __construct(int $orderDetailId)
```

jolelievre commented on src/Core/Domain/Order/ValueObject/OrderDetailId.php:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * @return int
+     */
+    public function getValue()
```
```suggestion
    public function getValue(): int
```

jolelievre commented on src/Core/Domain/Order/ValueObject/OrderDetailId.php:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * @param int $orderId
+     */
+    private function assertIntegerIsGreaterThanZero($orderId)
```
```suggestion
    private function assertIntegerIsGreaterThanZero(int $orderId): void
```

jolelievre commented on src/Core/Domain/Order/ValueObject/OrderDetailId.php:
```diff
@@ -0,0 +1,70 @@
+     */
+    private function assertIntegerIsGreaterThanZero($orderId)
+    {
+        if (!is_int($orderId) || 0 > $orderId) {
```
```suggestion
        if (0 > $orderId) {
```

jolelievre replied:
Always try to force the type when you can 😉

jolelievre commented on src/Core/Domain/OrderReturn/Command/BulkDeleteProductFromOrderReturnCommand.php:
```diff
@@ -0,0 +1,93 @@
+
+    /**
+     * @param int $orderReturnId
+     * @param array $orderReturnDetailIds
```
```suggestion
     * @param int[] $orderReturnDetailIds
```
Can you specify that an array of int is expected please?

jolelievre commented on src/Core/Domain/OrderReturn/Command/BulkDeleteProductFromOrderReturnCommand.php:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    /**
+     * @param array $orderReturnDetailIds
```
```suggestion
     * @param int[] $orderReturnDetailIds
```

jolelievre commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -86,4 +110,122 @@ public function update(OrderReturn $orderReturn): void
+     * @throws OrderReturnException
+     */
+    public function deleteOrderReturnProduct(
+        OrderReturnId $orderReturnId,
```
Do you really need to specify the OrderReturnId? From what I see in the DB you can get it from the DB since the detail is associated to an order return This would make this method and the command parameters simpler

jolelievre replied:
You can add a `public function getOrderReturnIdByOrderDetailId(int $orderDetailId): int` method in this repository

JevgenijVisockij replied:
The sole reason I added orderReturnId is because legacy method also uses it, but if it's refactored as you suggested in next comment, then yeah it's not needed.

jolelievre commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -86,4 +110,122 @@ public function update(OrderReturn $orderReturn): void
+     * @throws \Doctrine\DBAL\Driver\Exception
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function getOrderReturnDetailByOrderDetailId(int $orderDetailId): OrderReturnDetail
```
Hmm does it mean the `id_order_detail` from the `order_return_detail` table is actually a reference to the `id_order_detail` from the `order_detail` table?

JevgenijVisockij replied:
yes

jolelievre commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -86,4 +110,122 @@ public function update(OrderReturn $orderReturn): void
+                $orderReturnId->getValue(),
+                $orderReturnDetailId->getValue(),
+                $customizationId ? $customizationId->getValue() : 0
+            )) {
```
Considering the limited thing this legacy method does (only one ligne DB query), I think it's worth replacing it with a query executed via DBAL connection (especially since it's already injected in this repository)

JevgenijVisockij replied:
Refactored to use connection

jolelievre commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -86,4 +110,122 @@ public function update(OrderReturn $orderReturn): void
+     *
+     * @throws OrderReturnException
+     */
+    public function getOrderReturnDetails(OrderReturnId $orderReturnId, Order $order): array
```
```suggestion
    public function getOrderReturnDetails(OrderReturnId $orderReturnId): array
```
Considering how the `$order` is used in this legacy method `OrderReturn::getOrdersReturnProducts` I agree it may be a bit too long to refacto for now. However since the `$order` is never used in the handler that calls this method I think the order fetching should be done here, instead of inject the `OrderRepository` in the handle you should inject it here and get the Order object at the beginning of this method (you can get the order ID via a DBAL query or by fetch the OrderReturn object as you prefer)

JevgenijVisockij replied:
Makes sense, refactored it

jolelievre commented on src/Core/Domain/OrderReturn/Command/BulkDeleteProductFromOrderReturnCommand.php:
```diff
@@ -0,0 +1,93 @@
+     * @throws OrderReturnException
+     */
+    public function __construct(
+        int $orderReturnId,
```
Do you really need the orderReturnId? Since it can be deduced by the `orderReturnDetailIds` anyway?

JevgenijVisockij replied:
Here ir removal is not needed since orderReturnId is needed to create orderReturnDetaiId

jolelievre commented on src/Core/Domain/OrderReturn/Command/DeleteProductFromOrderReturnCommand.php:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * @var OrderReturnId
+     */
+    private $orderReturnId;
```
Same suggestion here, not sure the `orderReturnId` is really necessary

jolelievre replied:
Oh, wait! Unless it's possible to have multiple OrderReturn linked to the same OrderDetail? Then indeed it would be mandatory everywhere, but then you can remove the `OrderReturnDetailId` VO because it means simply this ID is not enough to identify a row in `order_return_detail` table, or rather you should transform it into a VO that needs to parameters `OrderReturnDetailId(orderReturnId, orderDetailId)`

JevgenijVisockij replied:
Yeah it's possible to do that, okay will refactor it to include orderReturnDetaiId

jolelievre commented on src/Core/Domain/OrderReturn/CommandHandler/DeleteProductFromOrderReturnHandler.php:
```diff
@@ -0,0 +1,66 @@
+                $command->getOrderReturnId(),
+                $command->getOrderReturnDetailId()
+            );
+        } catch (OrderReturnException $e) {
```
I don't think you need to try/catch here since the most likely exception to be thrown by `deleteOrderReturnProduct` is already a `DeleteOrderReturnProductException` So you can just call the repository method right away

JevgenijVisockij replied:
Especially since i refactored deleteOrderReturnDetail to not use legacy I agree. Removed Exception catching.

jolelievre commented on src/Core/Domain/OrderReturn/CommandHandler/BulkDeleteProductFromOrderReturnHandler.php:
```diff
@@ -0,0 +1,130 @@
+            if (isset($details[$orderReturnDetailId->getValue()])) {
+                unset($details[$orderReturnDetailId->getValue()]);
+            } else {
+                $errors[] = $orderReturnDetailId->getValue();
```
```suggestion
                $unknownDetails[] = $orderReturnDetailId->getValue();
```

jolelievre commented on src/PrestaShopBundle/EventListener/FilterOrderReturnProductsSearchCriteriaListener.php:
```diff
@@ -0,0 +1,101 @@
+ * Class FilterOrderReturnProductsSearchCriteriaListener is responsible for updating OrderReturnProducts filter with
+ * order return product id.
+ */
+class FilterOrderReturnProductsSearchCriteriaListener
```
Instead of this listener I recommend you to inspire from this filter builder https://github.com/PrestaShop/PrestaShop/blob/c583dc50bb39f7a808fecfa18cfa7192f39d30b4/src/Core/Search/Builder/TypedBuilder/ProductCombinationFiltersBuilder.php#L43

It's the same principle you need the parent ID to be in the filters/searchcriteria values, but using `TypedFiltersBuilderInterface` is clear it's part of the parameter resolver feature, and you can unit test your custom builder more easily. It also enforces the fact that `OrderReturnProductsFilters` is not valid unless it has the parent ID defined

JevgenijVisockij replied:
I don't see this being used at all, I guess at some point I implemented using this and then forgot to remove it. 
Because filtering works without this and there is no search(I think it's not really needed).

Removed it

jolelievre commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml:
```diff
@@ -43,8 +47,57 @@ admin_order_returns_update:
     orderReturnId: \d+
+
+admin_order_returns_products_filter:
+  path: /{orderReturnId}/edit
```
Be careful you may have conflicts with `admin_order_returns_edit` which has the same path and also handles the POST method, if the POST request always has to pass through this route then you should remove the method from the `admin_order_returns_edit` route

Also it's a detail but if you can place this route right under `admin_order_returns_edit` it will be easier to read/understand

JevgenijVisockij replied:
Good point! Changed edit to only use GET

jolelievre commented on src/PrestaShopBundle/Resources/config/services/adapter/order_return.yml:
```diff
@@ -21,10 +21,47 @@ services:
+      - '@prestashop.adapter.order_return.validator.order_return_validator'
+
+  prestashop.adapter.order_return.repository.order_return_repository:
+    alias: PrestaShop\PrestaShop\Adapter\OrderReturn\Repository\OrderReturnRepository
```
```suggestion
    alias: PrestaShop\PrestaShop\Adapter\OrderReturn\Repository\OrderReturnRepository
    deprecated: ~
```

When you define aliases like this you should also add the deprecated config, this way we'll be able to remove them in a next major

jolelievre commented on src/PrestaShopBundle/Resources/config/services/adapter/order.yml:
```diff
@@ -284,5 +284,8 @@ services:
+
   prestashop.adapter.order.repository.order_repository:
-    class: PrestaShop\PrestaShop\Adapter\Order\Repository\OrderRepository
+    alias: PrestaShop\PrestaShop\Adapter\Order\Repository\OrderRepository
```
```suggestion
    alias: PrestaShop\PrestaShop\Adapter\Order\Repository\OrderRepository
    deprecated: ~
```

jolelievre commented on src/Core/Domain/OrderReturn/OrderReturnSettings.php:
```diff
@@ -0,0 +1,34 @@
+
+class OrderReturnSettings
+{
+    public const ORDER_RETURN_STATE_WAITING_FOR_PACKAGE_ID = 2;
```
Maybe you should complete this with the other states even if they are all integers and even if they are not used yet in the recent handler But at least it will allow identifying the different steps/states

Ideally you could also replace the hard-coded values in the legacy code with these consts

jolelievre replied:
Oh wait.. I just realized that this is actually an ID from the DB based on the fixtures... And we even use some checks like `if ($orderReturn->state < OrderReturnSettings::ORDER_RETURN_STATE_WAITING_FOR_PACKAGE_ID)` So basically our domain rules are based on the assumption that the order states were added in a specific order and never modified since Even though they are editable 🤦‍♂️

jolelievre replied:
I know it's not your fault and this is all from the legacy, but I really don't think we should continue handling this feature this way. We need to see with @MatShir how the return states are supposed to behave and improve them while we are doing this migration

MatShir replied:
I suggest keeping the functional workflow with in these states as it is. But I think like @jolelievre that this is an opportunity to make it more robust and coherent.

JevgenijVisockij replied:
Moved PDF and order return logic to separate PR
https://github.com/PrestaShop/PrestaShop/pull/32048

jolelievre reviewed (commented):
Thanks @JevgenijVisockij 

First review 😅 Lots of comments but most important I discovered how the return state are handled based on hard-coded assumed IDs from the database and I don't think we should keep using those as references

So we need to think of a new better way to handle these states

JevgenijVisockij committed:
Apply suggestions from code review

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>
Co-authored-by: Julius Žukauskas <31609858+zuk3975@users.noreply.github.com>
Co-authored-by: Daniel Hlavacek <daniel.hlavacek@hotmail.cz>

JevgenijVisockij committed:
Update src/Core/Domain/OrderReturn/QueryResult/OrderDetailCustomizations.php

Changing addOrderDetailCustomization to setOrderDetailCustomizations

Co-authored-by: Julius Žukauskas <31609858+zuk3975@users.noreply.github.com>

JevgenijVisockij committed:
Refactored orderReturnDetailId to include order_return_id as it's needed. Changed delet order return detail to not use legacy

JevgenijVisockij committed:
Fixed bulk deletion not working

JevgenijVisockij committed:
Removed exception catching in DeleteProductFromOrderReturnHandler.php

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Removing PDF printing from this PR

JevgenijVisockij committed:
Removed uneeded function

JevgenijVisockij committed:
Removed uneeded file

JevgenijVisockij committed:
Removed unneeded divs

JevgenijVisockij committed:
Removing unneeded use

JevgenijVisockij committed:
Merge pull request #31996 from Progi1984/removeDeprecate019

Remove triggered errors and deprecated for Grid

zuk3975 commented on src/Core/Domain/Order/ValueObject/OrderDetailId.php:
```diff
@@ -0,0 +1,70 @@
+    private function assertIntegerIsGreaterThanZero(int $orderId): void
+    {
+        if (0 > $orderId) {
+            throw new OrderException(sprintf('Order detail id %s is invalid. Order detail id must be number that is greater than zero.', var_export($orderId, true)));
```
var export is not needed. It was used some time ago when we didn't have strict types. Now you know it is of type int, so no point of that anymore.
```suggestion
            throw new OrderException(sprintf('Order detail id %d is invalid. Order detail id must be number that is greater than zero.', $orderId));
```
Not blocking though

zuk3975 commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -71,4 +77,37 @@ public function get(OrderId $orderId): Order
+     *
+     * @return OrderDetailCustomizations|null
+     *
+     * @throws PrestaShopException
+     * @throws \PrestaShopDatabaseException
```
shouldn't be thrown.
Repository should wrap the old code and don't let the legacy exceptions to bubble up, but instead throw CoreException (check examples in abstract repository)

zuk3975 commented on src/Adapter/OrderReturn/QueryHandler/GetOrderDetailCustomizationsHandler.php:
```diff
@@ -0,0 +1,70 @@
+     */
+    private $contextLangId;
+
+    /** @var OrderRepository */
```
```suggestion
    /**
     * @var OrderRepository
     */

zuk3975 commented on src/Core/Grid/Filter/OrderReturnProductsFilterFormFactory.php:
```diff
@@ -0,0 +1,108 @@
+            throw new OrderReturnException('This page needs to have orderReturnId as a parameter');
+        }
+
+        $queryParams['orderReturnId'] = $request->attributes->get('orderReturnId');
```
if followed example above with custom Grid factory, then this probably might not be needed anymore (so this whole filter form factory decorator could probably be removed), as all filters should be passed already to searchCriteria (when building the searchCriteria from request with your custom FiltersBuilder)

JevgenijVisockij replied:
Yeah you are right removed it

zuk3975 commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml:
```diff
@@ -23,28 +25,69 @@ admin_merchandise_returns_filter:
   requirements:
     orderReturnId: \d+
 
+admin_order_returns_products_filter:
```
not a big deal, but I think we are always referring as `search` and not `filter`? (+ the action is called searchGrid, so in theory it should be `...search_grid:`, but search should be ok as well, as there are lots usages like this :man_shrugging: 
```suggestion
admin_order_return_products_search:
```

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/adapter/order_return_state.yml:
```diff
@@ -20,8 +20,10 @@ services:
 
+  PrestaShop\PrestaShop\Adapter\OrderReturnState\Repository\OrderReturnStateRepository:
+
   prestashop.adapter.order_return_state.repository.order_return_state_repository:
```
deprecate?

zuk3975 reviewed (dismissed):
There are some change requests that would be nice to fix, but IMO they could be considered improvements and fixed in dedicated PR considering this one used older practices that we used before (and by those older standards it should be ok. It is big and it has been in a feedback loop for a long time already)

JevgenijVisockij committed:
Added exception catching for repository

JevgenijVisockij committed:
Update src/Adapter/OrderReturn/QueryHandler/GetOrderDetailCustomizationsHandler.php

Code style fix

Co-authored-by: Julius Žukauskas <31609858+zuk3975@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml

Renaming of route

Co-authored-by: Julius Žukauskas <31609858+zuk3975@users.noreply.github.com>

JevgenijVisockij committed:
Added deprecation and fixed route

JevgenijVisockij committed:
Update src/Core/Domain/Order/ValueObject/OrderDetailId.php

Removing uneeded var_export

Co-authored-by: Julius Žukauskas <31609858+zuk3975@users.noreply.github.com>

FabienPapet commented on src/Core/Domain/OrderReturn/QueryResult/OrderDetailCustomizations.php:
```diff
@@ -0,0 +1,65 @@
+     */
+    private function setOrderDetailCustomizations(array $orderDetailCustomizations): void
+    {
+        foreach ($orderDetailCustomizations as $orderDetailCustomization) {
+            $this->orderDetailCustomizations[] = $orderDetailCustomization;
+        }
```
```suggestion
$this->orderDetailCustomizations = orderDetailCustomizations
```

JevgenijVisockij committed:
REmoving dump

JevgenijVisockij committed:
Removed no longer needed OrderReturnProductsFilterFormFactory.php

JevgenijVisockij committed:
CS fixes

# Conflicts:
#	classes/lang/KeysReference/FeatureFlagLang.php

JevgenijVisockij committed:
Clarified routes for order return

# Conflicts:
#	src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml

JevgenijVisockij committed:
Changes so  lang id doesn't come from context but form arugments

JevgenijVisockij committed:
Cs fixer fixes

JevgenijVisockij committed:
Fixed php stan not passing

JevgenijVisockij committed:
Added needed space

nicosomb commented:
Just a question about the title of this PR: is that clear enough with this title in the changelog?

JevgenijVisockij commented:
> Just a question about the title of this PR: is that clear enough with this title in the changelog?

Hey, updated the title to reflect what this PR focuses on better.

JevgenijVisockij committed:
refactored for cleaner object in order return id

# Conflicts:
#	src/Core/Grid/GridFactory.php
#	src/PrestaShopBundle/Resources/config/services/core/search.yml

JevgenijVisockij committed:
Fixed issue with grid interface

JevgenijVisockij committed:
CS fixer fix

JevgenijVisockij commented:
PR is ready for review, once this and https://github.com/PrestaShop/PrestaShop/pull/32048 are merged only things needed to finish merchandise return migration is to rename all remaining classes in code to OrderReturn from MerchandiseReturn and remove feature flag.
